&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&ensp;
[Home Page](../../START_HERE.ipynb)

[Previous Notebook](Challenge.ipynb)
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
[1](Challenge.ipynb)
[2]


# Solution
Scikit-Learn is an incredibly powerful toolkit that allows data scientists to quickly build models from their data, and it one of the most common and useful tools in the Python data science ecosystem. cuML is the RAPIDS library that implements similar machine learning algorithms that use CUDA to run on GPUs, with an API that mirrors the Scikit-learn one as much as possible.

This dataset comes from a proof-of-concept study published in 1999 by Golub et al. It showed how new cases of cancer could be classified by gene expression monitoring (via DNA microarray) and thereby provided a general approach for identifying new cancer classes and assigning tumors to known classes. These data were used to classify patients with acute myeloid leukemia (AML) and acute lymphoblastic leukemia (ALL).

Here is the dataset link: https://www.kaggle.com/crawford/gene-expression

## Here is the list of exercises and modules to work on in the lab:

- Convert the serial Pandas computations to CuDF operations.
- Utilize CuML to accelerate the machine learning models.
- Experiment with Dask to create a cluster and distribute the data and scale the operations.

You will start writing code from <a href='#ex'>here</a>, but make sure you execute the data processing blocks to understand the dataset.



The first step is downloading the dataset and putting it in the data directory, for using in this tutorial. Download the dataset here, and place it in (host/data) folder. Now we will import the necessary libraries.

In [1]:
import matplotlib.pyplot as plt
import numpy as np; print('NumPy Version:', np.__version__)
%matplotlib inline
import sys
import sklearn; print('Scikit-Learn Version:', sklearn.__version__)
from sklearn.linear_model import LinearRegression
from sklearn import preprocessing 
import pandas as pd
from sklearn.utils import resample
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.feature_selection import SelectFromModel
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, roc_curve, auc
from sklearn.preprocessing import OrdinalEncoder, StandardScaler
import cudf
import cupy
import matplotlib.pyplot as plt
# import for model building
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from cuml.linear_model import MBSGDRegressor as cumlSGD
from sklearn.linear_model import SGDRegressor as skSGD
from sklearn.datasets import make_regression
from sklearn.metrics import mean_squared_error
from cuml.ensemble import RandomForestClassifier as curfc
from sklearn.ensemble import RandomForestClassifier as skrfc
from cuml import make_regression
from cuml.linear_model import LinearRegression as cuLinearRegression
from cuml.metrics.regression import r2_score
from sklearn.linear_model import LinearRegression as skLinearRegression
from cuml.neighbors import KNeighborsClassifier as KNeighborsC
from sklearn.neighbors import KNeighborsClassifier
from cuml.linear_model import MBSGDClassifier as cumlMBSGDClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from cuml import Ridge
from cuml.linear_model import Ridge
from sklearn.linear_model import Ridge
from cuml import LogisticRegression
from sklearn.linear_model import LogisticRegression as skLogistic
from cuml.linear_model import ElasticNet
from sklearn import linear_model
from cuml.linear_model import Lasso
from cuml.solvers import SGD as cumlSGD
from sklearn.metrics import accuracy_score
from sklearn import model_selection, datasets
from cuml.dask.common import utils as dask_utils
from dask.distributed import Client, wait
from dask_cuda import LocalCUDACluster
import dask_cudf

from cuml.dask.ensemble import RandomForestClassifier as cumlDaskRF
from sklearn.ensemble import RandomForestClassifier as sklRF

NumPy Version: 1.19.2
Scikit-Learn Version: 0.23.1


We'll read the dataframe into y from the csv file, view its dimensions and observe the first 5 rows of the dataframe.

In [2]:
%%time
y = pd.read_csv('../../data/actual.csv')
print(y.shape)
y.head()

(72, 2)
CPU times: user 2.11 ms, sys: 2.15 ms, total: 4.27 ms
Wall time: 3.95 ms


patient cancer
0        1    ALL
1        2    ALL
2        3    ALL
3        4    ALL
4        5    ALL

Let's convert our target variable categories to numbers.

In [3]:
y['cancer'].value_counts()
# Recode label to numeric
y = y.replace({'ALL':0,'AML':1})
labels = ['ALL', 'AML'] # for plotting convenience later on

Read the training and test data provided in the challenge from the data folder. View their dimensions.

In [4]:
# Import training data
df_train = pd.read_csv('../../data/data_set_ALL_AML_train.csv')
print(df_train.shape)

# Import testing data
df_test = pd.read_csv('../../data/data_set_ALL_AML_independent.csv')
print(df_test.shape)

(7129, 78)
(7129, 70)


Observe the first few rows of the train dataframe and the data format.

In [5]:
df_train.head()

Gene Description Gene Accession Number    1 call    2  \
0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -214    A -139   
1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -153    A  -73   
2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at  -58    A   -1   
3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at   88    A  283   
4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -295    A -264   

  call.1    3 call.2    4 call.3  ...   29 call.33   30 call.34   31 call.35  \
0      A  -76      A -135      A  ...   15       A -318       A  -32       A   
1      A  -49      A -114      A  ... -114       A -192       A  -49       A   
2      A -307      A  265      A  ...    2       A  -95       A   49       A   
3      A  309      A   12      A  ...  193       A  312       A  230       P   
4      A -376      A -419      A  ...  -51       A -139       A -367       A   

    32 call.36   33 call.37  
0 -124       A -135       A  
1  -79       A -186       A  
2  -37       A  -70       A  
3  330       A  337       A  
4 -188       A -407       A  

[5 rows x 78 columns]

Observe the first few rows of the test dataframe and the data format.

In [6]:
df_test.head()

Gene Description Gene Accession Number   39 call   40  \
0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -342    A  -87   
1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -200    A -248   
2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   41    A  262   
3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at  328    A  295   
4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -224    A -226   

  call.1   42 call.2   47 call.3  ...   65 call.29   66 call.30   63 call.31  \
0      A   22      A -243      A  ...  -62       A  -58       A -161       A   
1      A -153      A -218      A  ... -198       A -217       A -215       A   
2      A   17      A -163      A  ...   -5       A   63       A  -46       A   
3      A  276      A  182      A  ...  141       A   95       A  146       A   
4      A -211      A -289      A  ... -256       A -191       A -172       A   

    64 call.32   62 call.33  
0  -48       A -176       A  
1 -531       A -284       A  
2 -124       A  -81       A  
3  431       A    9       A  
4 -496       A -294       A  

[5 rows x 70 columns]

As we can see, the data set has categorical values but only for the columns starting with "call". We won't use the columns having categorical values, but remove them.

In [7]:
# Remove "call" columns from training and testing data
train_to_keep = [col for col in df_train.columns if "call" not in col]
test_to_keep = [col for col in df_test.columns if "call" not in col]

X_train_tr = df_train[train_to_keep]
X_test_tr = df_test[test_to_keep]

Rename the columns and reindex for formatting purposes and ease in reading the data.

In [8]:
train_columns_titles = ['Gene Description', 'Gene Accession Number', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
       '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', 
       '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38']

X_train_tr = X_train_tr.reindex(columns=train_columns_titles)

In [9]:
test_columns_titles = ['Gene Description', 'Gene Accession Number','39', '40', '41', '42', '43', '44', '45', '46',
       '47', '48', '49', '50', '51', '52', '53',  '54', '55', '56', '57', '58', '59',
       '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72']

X_test_tr = X_test_tr.reindex(columns=test_columns_titles)

We will take the transpose of the dataframe so that each row is a patient and each column is a gene.

In [10]:
X_train = X_train_tr.T
X_test = X_test_tr.T

print(X_train.shape) 
X_train.head()

(40, 7129)


0     \
Gene Description       AFFX-BioB-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-5_at   
1                                                     -214   
2                                                     -139   
3                                                      -76   

                                                      1     \
Gene Description       AFFX-BioB-M_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-M_at   
1                                                     -153   
2                                                      -73   
3                                                      -49   

                                                      2     \
Gene Description       AFFX-BioB-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-3_at   
1                                                      -58   
2                                                       -1   
3                                                     -307   

                                                      3     \
Gene Description       AFFX-BioC-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioC-5_at   
1                                                       88   
2                                                      283   
3                                                      309   

                                                      4     \
Gene Description       AFFX-BioC-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioC-3_at   
1                                                     -295   
2                                                     -264   
3                                                     -376   

                                                       5     \
Gene Description       AFFX-BioDn-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioDn-5_at   
1                                                      -558   
2                                                      -400   
3                                                      -650   

                                                       6     \
Gene Description       AFFX-BioDn-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioDn-3_at   
1                                                       199   
2                                                      -330   
3                                                        33   

                                                      7     \
Gene Description       AFFX-CreX-5_at (endogenous control)   
Gene Accession Number                       AFFX-CreX-5_at   
1                                                     -176   
2                                                     -168   
3                                                     -367   

                                                      8     \
Gene Description       AFFX-CreX-3_at (endogenous control)   
Gene Accession Number                       AFFX-CreX-3_at   
1                                                      252   
2                                                      101   
3                                                      206   

                                                      9     ...  \
Gene Description       AFFX-BioB-5_st (endogenous control)  ...   
Gene Accession Number                       AFFX-BioB-5_st  ...   
1                                                      206  ...   
2                                                       74  ...   
3                                                     -215  ...   

                                                            7119  \
Gene Description       Transcription factor Stat5b (stat5b) mRNA   
Gene Accession Number                                  U48730_at   
1                                                            185   
2         

Just clearning the data, removing extra columns and converting to numerical values.

In [11]:
# Clean up the column names for training and testing data
X_train.columns = X_train.iloc[1]
X_train = X_train.drop(["Gene Description", "Gene Accession Number"]).apply(pd.to_numeric)

# Clean up the column names for Testing data
X_test.columns = X_test.iloc[1]
X_test = X_test.drop(["Gene Description", "Gene Accession Number"]).apply(pd.to_numeric)

print(X_train.shape)
print(X_test.shape)
X_train.head()

(38, 7129)
(34, 7129)


Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \
1                                -214            -153             -58   
2                                -139             -73              -1   
3                                 -76             -49            -307   
4                                -135            -114             265   
5                                -106            -125             -76   

Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \
1                                  88            -295             -558   
2                                 283            -264             -400   
3                                 309            -376             -650   
4                                  12            -419             -585   
5                                 168            -230             -284   

Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \
1                                  199            -176             252   
2                                 -330            -168             101   
3                                   33            -367             206   
4                                  158            -253              49   
5                                    4            -122              70   

Gene Accession Number  AFFX-BioB-5_st  ...  U48730_at  U58516_at  U73738_at  \
1                                 206  ...        185        511       -125   
2                                  74  ...        169        837        -36   
3                                -215  ...        315       1199         33   
4                                  31  ...        240        835        218   
5                                 252  ...        156        649         57   

Gene Accession Number  X06956_at  X16699_at  X83863_at  Z17240_at  \
1                            389        -37        793        329   
2                            442        -17        782        295   
3                            168         52       1138        777   
4                            174       -110        627        170   
5                            504        -26        250        314   

Gene Accession Number  L49218_f_at  M71243_f_at  Z78285_f_at  
1                               36          191          -37  
2                               11           76          -14  
3                               41          228          -41  
4                              -50          126          -91  
5                               14           56          -25  

[5 rows x 7129 columns]

We have the 38 patients as rows in the training set, and the other 34 as rows in the testing set. Each of those datasets has 7129 gene expression features. But we haven't yet associated the target labels with the right patients. You will recall that all the labels are all stored in a single dataframe. Let's split the data so that the patients and labels match up across the training and testing dataframes.We are now splitting the data into train and test sets. We will subset the first 38 patient's cancer types.

In [12]:
X_train = X_train.reset_index(drop=True)
y_train = y[y.patient <= 38].reset_index(drop=True)

# Subset the rest for testing
X_test = X_test.reset_index(drop=True)
y_test = y[y.patient > 38].reset_index(drop=True)

Generate descriptive statistics to analyse the data further.

In [13]:
X_train.describe()

Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \
count                       38.000000       38.000000       38.000000   
mean                      -120.868421     -150.526316      -17.157895   
std                        109.555656       75.734507      117.686144   
min                       -476.000000     -327.000000     -307.000000   
25%                       -138.750000     -205.000000      -83.250000   
50%                       -106.500000     -141.500000      -43.500000   
75%                        -68.250000      -94.750000       47.250000   
max                         17.000000      -20.000000      265.000000   

Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \
count                       38.000000       38.000000        38.000000   
mean                       181.394737     -276.552632      -439.210526   
std                        117.468004      111.004431       135.458412   
min                        -36.000000     -541.000000      -790.000000   
25%                         81.250000     -374.250000      -547.000000   
50%                        200.000000     -263.000000      -426.500000   
75%                        279.250000     -188.750000      -344.750000   
max                        392.000000      -51.000000      -155.000000   

Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \
count                        38.000000       38.000000       38.000000   
mean                        -43.578947     -201.184211       99.052632   
std                         219.482393       90.838989       83.178397   
min                        -479.000000     -463.000000      -82.000000   
25%                        -169.000000     -239.250000       36.000000   
50%                         -33.500000     -185.500000       99.500000   
75%                          79.000000     -144.750000      152.250000   
max                         419.000000      -24.000000      283.000000   

Gene Accession Number  AFFX-BioB-5_st  ...   U48730_at    U58516_at  \
count                       38.000000  ...   38.000000    38.000000   
mean                       112.131579  ...  178.763158   750.842105   
std                        211.815597  ...   84.826830   298.008392   
min                       -215.000000  ...   30.000000   224.000000   
25%                        -47.000000  ...  120.000000   575.500000   
50%                         70.500000  ...  174.500000   700.000000   
75%                        242.750000  ...  231.750000   969.500000   
max                        561.000000  ...  356.000000  1653.000000   

Gene Accession Number   U73738_at    X06956_at   X16699_at    X83863_at  \
count                   38.000000    38.000000   38.000000    38.000000   
mean                     8.815789   399.131579  -20.052632   869.052632   
std                     77.108507   469.579868   42.346031   482.366461   
min                   -178.000000    36.000000 -112.000000   195.000000   
25%                    -42.750000   174.500000  -48.000000   595.250000   
50%                     10.500000   266.000000  -18.000000   744.500000   
75%                     57.000000   451.750000    9.250000  1112.000000   
max                    218.000000  2527.000000   52.000000  2315.000000   

Gene Accession Number    Z17240_at  L49218_f_at  M71243_f_at  Z78285_f_at  
count                    38.000000    38.000000    38.000000    38.000000  
mean                    335.842105    19.210526   504.394737   -29.210526  
std                     209.826766    31.158841   728.744405    30.851132  
min                      41.000000   -50.000000    -2.000000   -94.000000  
25%                     232.750000     8.000000   136.000000   -42.750000  
50%                     308.500000    20.000000   243.500000   -26.000000  
75%                     389.500000    30.250000   487.250000   -11.500000  
max                    1109.000000   115.000000  3193.000000    36.000000  

[8 rows

Clearly there is some variation in the scales across the different features. Many machine learning models work much better with data that's on the same scale, so let's create a scaled version of the dataset.

In [14]:
X_train_fl = X_train.astype(float, 64)
X_test_fl = X_test.astype(float, 64)

# Apply the same scaling to both datasets
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train_fl)
X_test = scaler.transform(X_test_fl) # note that we transform rather than fit_transform

<a id='ex2'></a><br>
Convert the pandas dataframes to CuDF dataframes to carry out the further CuML tasks.

In [15]:
%%time
X_cudf_train = cudf.DataFrame(X_train)
X_cudf_test = cudf.DataFrame(X_test)

y_cudf_train = cudf.DataFrame(y_train)
#y_cudf_test = cudf.Series(y_test.values)

CPU times: user 1.58 s, sys: 435 ms, total: 2.02 s
Wall time: 2.03 s


Given below is the scikit-learn implementation for predicting the target variable using ElasticNet Classifier to show you how we can use this dataset and fit the model. Observe which dataframe and been used, and how the model is created.

## ElasticNet Classifier
### Scikit-learn model

#### Fit

In [16]:
%%time
regr = ElasticNet()
regr.fit(X_train, y_train.iloc[:,1])

CPU times: user 364 ms, sys: 77.4 ms, total: 441 ms
Wall time: 440 ms


ElasticNet(alpha=1.0, l1_ratio=0.5, fit_intercept=True, normalize=False, max_iter=1000, tol=0.001, selection='cyclic', handle=<cuml.raft.common.handle.Handle object at 0x7f13bc25ab10>, output_type='numpy', verbose=4)

#### Evaluate

In [17]:
%%time
X_test = X_test.astype(np.float64)
y_test = y_test.astype(np.float64)
print(regr.score(X_test,y_test.iloc[:,1]))

-0.06174317372378324
CPU times: user 1.91 ms, sys: 1.97 ms, total: 3.88 ms
Wall time: 3.11 ms


<a id='ex3'></a><br>

### CuML model

#### Fit

In [18]:
%%time
enet = ElasticNet()

enet.fit(X_cudf_train, y_train.iloc[:,1])

CPU times: user 696 ms, sys: 21 ms, total: 717 ms
Wall time: 721 ms


ElasticNet(alpha=1.0, l1_ratio=0.5, fit_intercept=True, normalize=False, max_iter=1000, tol=0.001, selection='cyclic', handle=<cuml.raft.common.handle.Handle object at 0x7f13bc25a9b0>, output_type='cudf', verbose=4)

### Evaluate

In [19]:
%%time
X_cudf_test = X_cudf_test.astype(np.float64)
print(enet.score(X_cudf_test, y_test.iloc[:,1]))

-0.06174317372378324
CPU times: user 1.12 s, sys: 602 µs, total: 1.12 s
Wall time: 1.12 s


# Logistic Regression

## Scikit-learn

### Fit

In [20]:
%%time
clf = skLogistic()
clf.fit(X_train, y_train.iloc[:,1])


CPU times: user 2.25 s, sys: 4.74 s, total: 6.99 s
Wall time: 119 ms


LogisticRegression()

### Evaluate

In [21]:
%%time
print(clf.score(X_test, y_test.iloc[:,1]))

0.8235294117647058
CPU times: user 57.7 ms, sys: 84.8 ms, total: 143 ms
Wall time: 2.23 ms


## CuML

### Fit

In [22]:
%%time
reg = LogisticRegression()
reg.fit(X_cudf_train,y_cudf_train.iloc[:,1])

CPU times: user 2.34 s, sys: 4.21 s, total: 6.54 s
Wall time: 408 ms


LogisticRegression(penalty='l2', tol=0.0001, C=1.0, fit_intercept=True, max_iter=1000, linesearch_max_iter=50, verbose=4, l1_ratio=None, solver='qn', handle=<cuml.raft.common.handle.Handle object at 0x7f13b388cd70>, output_type='cudf')

### Evaluate

In [23]:
%%time
print(reg.score(X_cudf_test, y_test.iloc[:,1]))

0.8529411554336548
CPU times: user 756 ms, sys: 1.18 ms, total: 757 ms
Wall time: 757 ms


# Nearest Neighbours Classifier

## Scikit-learn

### Fit

In [24]:
%%time
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(X_train, y_train.iloc[:,1])

CPU times: user 6.71 ms, sys: 677 µs, total: 7.39 ms
Wall time: 6.47 ms


KNeighborsClassifier(n_neighbors=3)

### Evaluate

In [25]:
%%time
print(neigh.score(X_test, y_test.iloc[:,1]))

0.7058823529411765
CPU times: user 17.1 ms, sys: 2.11 ms, total: 19.2 ms
Wall time: 18.6 ms


## CuML

### Fit

In [26]:
%%time
knn = KNeighborsC(n_neighbors=10)
knn.fit(X_cudf_train, y_cudf_train.iloc[:,1])

CPU times: user 1.24 s, sys: 14 ms, total: 1.26 s
Wall time: 1.26 s


KNeighborsClassifier(weights='uniform')

### Evaluate

In [27]:
%%time
print(knn.score(X_cudf_test, y_test.iloc[:,1]))

0.6470588445663452
CPU times: user 1.24 s, sys: 133 ms, total: 1.38 s
Wall time: 1.38 s


# Dask Integration

We will try using the Random Forests Classifier  and implement using CuML and Dask.

# Start Dask cluster

In [28]:
# This will use all GPUs on the local host by default
cluster = LocalCUDACluster(threads_per_worker=1)
c = Client(cluster)

# Query the client for all connected workers
workers = c.has_what().keys()
n_workers = len(workers)
n_streams = 8 # Performance optimization

## Define Parameters

In addition to the number of examples, random forest fitting performance depends heavily on the number of columns in a dataset and (especially) on the maximum depth to which trees are allowed to grow. Lower `max_depth` values can greatly speed up fitting, though going too low may reduce accuracy.

In [29]:
# Random Forest building parameters
max_depth = 12
n_bins = 16
n_trees = 1000

## Distribute data to worker GPUs

In [30]:
X_train = X_train.astype(np.float32)
X_test = X_test.astype(np.float32)
y_train = y_train.astype(np.int32)
y_test = y_test.astype(np.int32)

In [31]:
n_partitions = n_workers

def distribute(X, y):
    # First convert to cudf (with real data, you would likely load in cuDF format to start)
    X_cudf = cudf.DataFrame.from_pandas(pd.DataFrame(X))
    y_cudf = cudf.Series(y)

    # Partition with Dask
    # In this case, each worker will train on 1/n_partitions fraction of the data
    X_dask = dask_cudf.from_cudf(X_cudf, npartitions=n_partitions)
    y_dask = dask_cudf.from_cudf(y_cudf, npartitions=n_partitions)

    # Persist to cache the data in active memory
    X_dask, y_dask = \
      dask_utils.persist_across_workers(c, [X_dask, y_dask], workers=workers)
    
    return X_dask, y_dask

X_train_dask, y_train_dask = distribute(X_train, y_train.iloc[:,1])
X_test_dask, y_test_dask = distribute(X_test, y_test.iloc[:,1])

# Create the  Scikit-learn model

Since a scikit-learn equivalent to the multi-node multi-GPU K-means in cuML doesn't exist, we will use Dask-ML's implementation for comparison.

In [38]:
%%time

# Use all avilable CPU cores
skl_model = sklRF(max_depth=max_depth, n_estimators=n_trees, n_jobs=-1)
skl_model.fit(X_train, y_train.iloc[:,1])

CPU times: user 2.39 s, sys: 757 ms, total: 3.14 s
Wall time: 2.17 s


RandomForestClassifier(max_depth=12, n_estimators=1000, n_jobs=-1)


## Train the distributed cuML model

In [39]:
%%time

cuml_model = cumlDaskRF(max_depth=max_depth, n_estimators=n_trees, n_bins=n_bins, n_streams=n_streams)
cuml_model.fit(X_train_dask, y_train_dask)

wait(cuml_model.rfs) # Allow asynchronous training tasks to finish

CPU times: user 66.5 ms, sys: 9.68 ms, total: 76.2 ms
Wall time: 1.06 s


DoneAndNotDoneFutures(done={<Future: finished, type: cuml.RandomForestClassifier, key: _construct_rf-2e8840d7-b3db-48fa-ab39-009b433976e5>, <Future: finished, type: cuml.RandomForestClassifier, key: _construct_rf-b2094e50-7d2c-49d9-9f28-444e287e85d2>}, not_done=set())

# Predict and check accuracy

In [40]:
skl_y_pred = skl_model.predict(X_test)
cuml_y_pred = cuml_model.predict(X_test_dask).compute().to_array()

# Due to randomness in the algorithm, you may see slight variation in accuracies
print("SKLearn accuracy:  ", accuracy_score(y_test.iloc[:,1], skl_y_pred))
print("CuML accuracy:     ", accuracy_score(y_test_dask.compute().to_array(), cuml_y_pred))

SKLearn accuracy:   0.7941176470588235
CuML accuracy:      0.5882352941176471


<a id='ex4'></a><br>

# CONCLUSION

Let's compare the performance of our solution!

| Algorithm     | Implementation | Accuracy      | Time | Algorithm     | Implementation | Accuracy      | Time |
| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |
| ElasticNet    | Scikit-learn       | -0.06      | 428 ms      | ElasticNet      | CuML      | -0.06      | 680 ms      |
| Logistic Regression     | Scikit-learn        | 0.82     | 173 ms       | Logistic Regression   | CuML       | 0.85     | 2.52 s       |
| Nearest Neighbours Classifier     | Scikit-learn       | 0.70      | 6.17 ms       | Nearest Neighbours Classifier     | CuML        | 0.64     | 1.74 s     |
| Random Forests Classifier  | Scikit-learn     | 0.72     |   2.17 s  | Random Forests Classifier | Dask-CuML | 0.64  | 1.06 s

Write down your observations and compare the CuML and Scikit learn scores. They should be approximately equal.  We hope that you found this exercise exciting and beneficial in understanding RAPIDS better. Share your highest accuracy and try to use the unique features of RAPIDS for accelerating your data science pipelines. Don't restrict yourself to the previously explained concepts, but use the documentation to apply more models and functions and achieve the best results.

##### Thus we can observe that for most cases, the CuML implementation is increasing the computation time by a few milliseconds, and time difference isnt seen as the data size is small in this sample. However as you increase the dataset, the same implementation shows drastic results. The Dask Implementation reduces the computation time by half and utilizes multiple GPUs to carry out the calculations. The slight difference in accuracy is due to the slight difference in runtime execution of both codes, but the accuracy is nearly constant. 



# References



<p xmlns:dct="http://purl.org/dc/terms/">
  <a rel="license"
     href="http://creativecommons.org/publicdomain/zero/1.0/">
    <center><img src="http://i.creativecommons.org/p/zero/1.0/88x31.png" style="border-style: none;" alt="CC0"  /></center>
  </a>
 
</p>


- The dataset is licensed under a CC0: Public Domain license.

- Molecular Classification of Cancer: Class Discovery and Class Prediction by Gene Expression. Science 286:531-537. (1999). Published: 1999.10.14. T.R. Golub, D.K. Slonim, P. Tamayo, C. Huard, M. Gaasenbeek, J.P. Mesirov, H. Coller, M. Loh, J.R. Downing, M.A. Caligiuri, C.D. Bloomfield, and E.S. Lander


## Licensing
  
This material is released by NVIDIA Corporation under the Creative Commons Attribution 4.0 International (CC BY 4.0).

[Previous Notebook](Challenge.ipynb)
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
[1](Challenge.ipynb)
[2]
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;


&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&ensp;
[Home Page](../../START_HERE.ipynb)